src/ImageRunModal.jsx

```diff
@@ -27,13 +27,112 @@ const units = {
     },
 };
 
+class PublishPortsForm extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            publish: [{ containerPort: null, hostPort: null, protocol: 'TCP' }],
+        };
+        this.removePort = this.removePort.bind(this);
+        this.addPort = this.addPort.bind(this);
+        this.onPortChange = this.onPortChange.bind(this);
+        this.validateAndChange = this.validateAndChange.bind(this);
+    }
+
+    removePort(idx, field, value) {
+        let ports = this.state.publish;
+        ports.splice(idx, 1);
+        if (ports.length === 0) {
+            this.addPort();
+        } else {
+            this.setState({ publish: ports });
```

Comment:
You might loose some updates here and in `addPort()` below. It's unlikely in this case, but it is preferred to use the functional version of `setState` in the cases were you need to update a state variable based on previous values. You pass a function which receives the actual previous state in an argument.

It's pretty well explained here: https://reactjs.org/docs/react-component.html#setstate

Reply:
Ah cool! I really should read that first.